Add documentation on traffic routing details (#9800)
istio/istio.io: content/en/docs/ops/configuration/traffic-management/traffic-routing/index.md

{{< /warning >}}

### ExternalName Services

An [ExternalName Service](https://kubernetes.io/docs/concepts/services-networking/service/#externalname) is essentially just a DNS alias.

Because there is no ClusterIP nor pod IPs to match on, for TCP ExternalName Services, all IPs on the port will be matched.
This may prevent [unmatched traffic](#unmatched-traffic) on the same port from being forwarded correctly.
As such, it is best to avoid these where possible, or use dedicated ports when needed.
HTTP and TLS do not share this constraint, as routing is done based on the hostname/SNI.

{{< warning >}}
Without Istio, the `ports` field of an ExternalName service is not required because the Service only represents a DNS entry.
However, with Istio the port must be declared in the Service, or it will [not be matched](#unmatched-traffic).
{{< /warning >}}

### ServiceEntry

In addition to Kubernetes Services, [ServiceEntries](/reference/config/networking/service-entry/#ServiceEntry) can be created to extend the set of services known to Istio.
This can be useful to ensure that traffic to external services, such as `example.com`, get the functionality of Istio.

> A ServiceEntry with `addresses` set will perform routing just like a ClusterIP Service.

Reviewer: I kind of get what this means but can you elaborate? If a SE has protocol TCP, then it needs to specify address for VIP matching else only ports are matched. Does resolution:DNS work in that case (I think it should) or users need to provide static IP addresses as endpoints?

Author: I will try to clarify a bit, but `resolution` and `address` are orthogonal. The resolution is just about upstream host selection, the addresses is for matching incoming traffic.

Reviewer: I meant to say that we should clarify how resolution DNS works with TCP..I understand address field (VIP) is about matching incoming traffic and endpoints is about where to send it if static..

Author: DNS resolution works the same with TCP. I am not sure why endpoints is relevant here since this is all about the routing part, probably I am missing the connection?

Reviewer: May be, a part of routing is eventually being able to select endpoints for the cluster.

---
title: Understanding Traffic Routing
linktitle: Traffic Routing
description: How Istio routes traffic through the mesh.
weight: 30
keywords: [traffic-management,proxy]
owner: istio/wg-networking-maintainers
test: n/a
---

> One of the goals of Istio is to act as a "transparent proxy" that can be dropped in place into an existing cluster and allow traffic to continue to function as previously.

Reviewer 2: ```suggestion
One of the goals of Istio is to act as a "transparent proxy" which can be dropped into an existing cluster, allowing traffic to continue to flow as before.
```

To understand what is happening in your mesh, it is important to understand how Istio routes traffic.

{{< warning >}}
This document describes low level implementation details. For a higher level overview, check out the traffic management [Concepts](/docs/concepts/traffic-management/) or [Tasks](/docs/tasks/traffic-management/).
{{< /warning >}}

## Protocols

Unlike Kubernetes, Istio has the ability to process higher level protocols such as HTTP and TLS.
In general, there are three classes of protocols Istio understands:

* HTTP, which includes HTTP/1.1, HTTP/2, and gRPC. Note that this does not include TLS encrypted traffic (HTTPS).
* TLS, which includes HTTPS.
* Opaque TCP bytes.

Which protocol is used is determined by [protocol selection](/docs/ops/configuration/traffic-management/protocol-selection/).

The use of "TCP" can be confusing, as in other contexts it is used to distinguish between other L4 protocols, such as UDP.
When referring to the TCP protocol in Istio, this typically means we are treating it as an opaque stream of bytes,
and not parsing higher level constructs such as TLS or HTTP.

## Traffic Routing

> When the Istio proxy receives a request, it must decide where to forward the request to, if anywhere.

Reviewer 2: ```suggestion
When an Envoy proxy receives a request, it must decide where, if anywhere, to forward it to.
```

I think we don't say "Istio proxy".  IANATW

> By default, this will be the original service that was request, unless [customized](/docs/tasks/traffic-management/traffic-shifting/).

Reviewer 2: ```suggestion
By default, this will be to the original service that was requested, unless [customized](/docs/tasks/traffic-management/traffic-shifting/).
```

that link is only to a page that talks about a certain use case; I think the others all apply too?  Maybe a link to the general concepts page?

> By default, this will be the original service that was request, unless [customized](/tasks/traffic-management/traffic-shifting/).

Author: I think this is very subtle, but the intent I had is:

* Service == Kubernetes Service object
* service == generic service. Could be from Service, ServiceEntry, etc

> By default, this will be the original service that was request, unless [customized](/tasks/traffic-management/traffic-shifting/).

Reviewer 2: "that was requested"

How this works depends on the protocol used.

### TCP

> When processing TCP traffic, Istio has a very small amount of useful information to route the connection based on - only the destination IP and Port.

Reviewer 3: ~~based on~~

> Fortunately, this is enough to determine the intended Service; the proxy is configured to listener on each service IP:Port pair and forward traffic to that service.

Reviewer 3: listen

For customizations, a TCP `VirtualService` can be configured, which allows [matching on specific IPs and ports](/docs/reference/config/networking/virtual-service/#L4MatchAttributes).

### TLS

> When processing TLS traffic, Istio has slightly more information available than raw TCP: we can inspect the [SNI](https://en.wikipedia.org/wiki/Server_Name_Indication).

Reviewer 2: SNI field?  SNI header?

Author: Based on the wiki article it seems this is how its used, although it does sound odd

Author: But neeraj said it as well so I will go with it

> For standard Services, the same IP:Port matching as raw TCP is used.

Reviewer 2: ```suggestion
For standard Services, the same IP:Port matching is used as for raw TCP.
```

> For standard Services, the same IP:Port matching as raw TCP is used.

Author: They may be out of cluster

> However, for services that do not have a Service IP defined, such as [ExternalName services](#externalname-services), the SNI will be used for routing.

Reviewer 4: standard Services == ClusterIP?

Additionally, custom routing can be configured with a TLS `VirtualService` to [matching on SNI](/config/networking/virtual-service/#TLSMatchAttributes).

### HTTP

HTTP allows much richer routing than TCP and TLS. With HTTP, we are able to route individual HTTP request, rather than just connections.
In addition, a [number of different request attributes](/config/networking/virtual-service/#HTTPMatchRequest) are available, such as host, path, headers, query parameters, etc.

While TCP and TLS traffic generally behave the same with or without Istio (assuming no configuration has been applied to customize the routing), HTTP has significant differences.

> * Istio will load balance individual requests. In general, this is highly desirably, especially in scenarios with long-lived connections such as gRPC and HTTP/2, where connection level load balancing is ineffectively.

Reviewer 3: ineffective.

> * Istio will load balance individual requests. In general, this is highly desirably, especially in scenarios with long-lived connections such as gRPC and HTTP/2, where connection level load balancing is ineffective.

Reviewer 4: highly desirable

> * Istio will load balance individual requests. In general, this is highly desirable, especially in scenarios with long-lived connections such as gRPC and HTTP/2, where connection level load balancing is ineffective.

Reviewer 5: only when sticky session is not used (which is not enabled by default)?

Author: Regardless of if consistent hash is used, its still load balancing individual requests technically - they just go to the same place. I don't want to make this complex here as a general concept doc, the API docs discuss the behavior of this feature

Reviewer 5: fair point - got it confused when you had long-lived connections.

> * Requests are routed based on the port and *`Host` header*, rather than port and IP. This means the destination IP address is effectively ignored. For example, `curl 8.8.8.8 -H "Host: productpage.default.svc.cluster.local"`, would be routed to the `productpage` Service.

Author: It is whatever the RFC says. If not, its a (security, likely) bug, not a docs issue :slightly_smiling_face:

> * Requests are routed based on the port and *`Host` header*, rather than port and IP. This means the destination IP address is effectively ignored. For example, `curl 8.8.8.8 -H "Host: productpage.default.svc.cluster.local"`, would be routed to the `productpage` Service.

Reviewer 5: pretty sure -H "Host: xx" works, as I use it a lot

## Unmatched traffic

> If traffic cannot be matched using one of the methods described above, it is treated as [passthrough traffic](/docs/tasks/traffic-management/egress/egress-control/#envoy-passthrough-to-external-services).

Author: Its *always* "passthrough" traffic. outbound policy determines what to do with passthrough traffic. 

That is how I see it anyways

> If traffic cannot be matched using one of the methods described above, it is treated as [passthrough traffic](/docs/tasks/traffic-management/egress/egress-control/#envoy-passthrough-to-external-services).

Reviewer 5: confused why outbound traffic policy is relevant here?  Isn't that only revenant when the destination is outside of the mesh/k8s cluster?

If service A -> B both in the mesh, would not it be a simple passthrough for any unmatched traffic?

> If traffic cannot be matched using one of the methods described above, it is treated as [passthrough traffic](/docs/tasks/traffic-management/egress/egress-control/#envoy-passthrough-to-external-services).

Author: "unmatched" is impossible to distinguish from "outside the mesh". If its unmatched, that means we know _nothing_ about the intended destination, so we have no clue where it is

> If traffic cannot be matched using one of the methods described above, it is treated as [passthrough traffic](/docs/tasks/traffic-management/egress/egress-control/#envoy-passthrough-to-external-services).

Reviewer 5: Ok, given "unmatched" could be in or out of the mesh, I like @howardjohn 's original texts better.  Adding an if condition could introduce confusion that unmatched is for external service.

> By default, these requests will be forwarded as-ism, which ensures that traffic to services that Istio is not aware of, such as external services that do not have `ServiceEntry`s created.

Reviewer 3: as-is

Reviewer 3: Sentence just seems to end. Maybe ... as-is, ensuring traffic to services that Istio is not aware of such as ..

Note that when these requests are forwarded, mutual TLS will not be used and telemetry collection is limited.

## Service types

Along with standard `ClusterIP` Services, Istio supports the full range of Kubernetes Services, with some caveats.

### `LoadBalancer` and `NodePort` Services

> These Services are supersets of `ClusterIP` Services, and are mostly concerned with allow access from external clients.

Reviewer 3: allowing
